tests/scripts/depends-ciphers.pl

```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env perl
+
+# depends-ciphers.pl
+#
+# Copyright (c) 2018, ARM Limited, All Rights Reserved
+#
+# Purpose
+#
+# To test the code dependencies on individual ciphers in each test suite. This
+# is a verification step to ensure we don't ship test suites that do not work
+# for some build options.
+#
+# The process is:
+#       for each possible cipher
+#           build the library and test suites with the hash disabled
+#           execute the test suites
+#
+# And any test suite with the wrong dependencies will fail.
+#
+# Usage: tests/scripts/depends-ciphers.pl
+#
+# This script should be executed from the root of the project directory.
+#
+# For best effect, run either with cmake disabled, or cmake enabled in a mode
+# that includes -Werror.
+
+use warnings;
+use strict;
+
+-d 'library' && -d 'include' && -d 'tests' or die "Must be run from root\n";
+
+my $config_h = 'include/mbedtls/config.h';
+
+# get a list of ciphers from the cipher module
+my $cipher_h = 'include/mbedtls/cipher.h';
+my $sed_cmd = 's/^    MBEDTLS_CIPHER_ID_\([A-Z0-9_]*\),.*/\1/p';
+my @ciphers = split( /\s+/, `sed -n -e '$sed_cmd' $cipher_h` );
+@ciphers = map { "MBEDTLS_${_}_C" } grep { ! m/NULL|3DES/ } @ciphers;
+
+# Some algorithms can't be disabled on their own as others depend on them, so
+# we list those reverse-dependencies here to keep check_config.h happy.
+my %revdeps = (
+    'MBEDTLS_AES_C'         => ['MBEDTLS_CTR_DRBG_C'],
```

Comment:
There are other dependencies on `MBEDTLS_AES_C`:
- `MBEDTLS_CMAC_C` ( also dependent on `MBEDTLS_DES_C`)
- `MBEDTLS_NIST_KW_C` 
- `MBEDTLS_GCM_C` ( also dependent on `MBEDTLS_CAMELLIA_C`)

We shouldn't rely on the default configuration.

Reply:
That's what I thought, and this is why I said that we shouldn't rely on the default configuration.

What if B is already undefined in the original config?

how about we start with a full configuration, and disable one by one?

Reply:
I agree that we should start from a full config.

Reply:
Am I missing something, or is it still starting from the default config?

I don't mind the script dealing with whatever `config.h` it's given, but the `all.sh` run seems to be the default.

Reply:
I don't think you're missing anything. This part is pending rework.